libs/gpu-codec/src/main/java/org/elasticsearch/gpu/codec/ES92GpuHnswSQVectorsFormat.java

```diff
@@ -69,13 +77,20 @@ public KnnVectorsWriter fieldsWriter(SegmentWriteState state) throws IOException
             state,
             maxConn,
             beamWidth,
-            flatVectorsFormat.fieldsWriter(state)
+            (FlatVectorsWriter) scalarQuantizedFormat.fieldsWriter(state)
```

Comment:
This cast indicates there's something a bit wonky in how `ES93ScalarQuantizedVectorsFormat` works. I think it's best if I have a closer look at those classes, and then I can integrate these changes in with any refactor of those classes I do